lib/Mojo/Redis/Protocol.pm

```diff
@@ -0,0 +1,240 @@
+package Mojo::Redis::Protocol;
+use Mojo::Base -base;
+
+use Carp qw(confess croak);
+use Encode 'find_encoding';
+
+use constant DEBUG => $ENV{MOJO_REDIS_DEBUG} || 0;
+
+sub encode {
+  my ($self, @stack) = @_;
+  my $encoder = $self->{encoder};
+
+  my $str = '';
+  while (@stack) {
+    my $message = shift @stack;
+    my $ref     = ref $message;
+    my $data    = $ref eq 'ARRAY' || !$ref ? $message : $message->{data};
+    my $type    = $ref eq 'ARRAY' ? '*' : !$ref ? '$' : $message->{type};
+
+    if ($type eq '+' or $type eq '-' or $type eq ':') {
+      $data = $encoder->encode($data, 0) if $encoder;
```

Comment:
None of these types should be encoded, because they are not null safe.

Reply:
What is the purpose of specifying 0 for encode and decode here?

Reply:
Note that if you specify any second argument for encode or decode, you should copy the string 
(like https://metacpan.org/source/SRI/Mojolicious-8.12/lib/Mojo/Util.pm#L134) because Encode will modify it in place.

Reply:
> None of these types should be encoded, because they are not binary safe.

What does "binary safe" mean? What will be the consequence of doing what I do here? Why should we not encode "+", but encode "$"?

> What is the purpose of specifying 0 for encode and decode here?

Just because I copy/pasted the code from Mojo::Util. Removed it now, since it's the default.

> Note that if you specify any second argument for encode or decode, you should copy the string

Ah! I forgot about that. Thanks. I'll change the code to replace the string in place.

Reply:
> > Note that if you specify any second argument for encode or decode, you should copy the string

I don't agree. I think that is only correct for the functions, not the methods. https://metacpan.org/pod/Encode::Encoding#-%3Eencode($string-[,$check])

Reply:
Binary safe means that bulk strings can contain any byte. Namely, simple strings, errors, and integers cannot contain the bytes `\r\n`, and there is no guarantee that encoding values in some encoding will result in this. I added a test for this.

The encode specification is a mess. What you linked is the documentation for how Encode subclasses maybe should work. But it's not how they work.